### ISSUE TITLE
OptString validation wants to be ok with an empty string, but Base doesn't

### ISSUE THREAD

**Issue by wchen-r7** [rapid7/metasploit-framework#10166]
OptString's validation code looks like this:

```ruby
  def valid?(value=self.value, check_empty: true)
    value = normalize(value)
    return false if check_empty && empty_required_value?(value)
    return super
  end
```

If you look at the code for empty_required_value?, it only checks for nil:

```ruby
    def empty_required_value?(value)
      required? && value.nil?
    end
```

However, when `valid?` calls `super`, this behavior is different:

```ruby
    def valid?(value, check_empty: true)
      if check_empty && required?
        # required variable not set
        return false if (value.nil? || value.to_s.empty?)
      end
      if regex
        return !!value.match(regex)
      end
      true
    end
```

As you can see, this time the value is being checked for empty. When there is a function that only cares about nil, and another one that cares both nil and an empty string, this behavior is not consistent, and it makes the design decision even more confusing, probably.

**Comment by asoto-r7** [comment]
@teebone reached out to us via Slack to report an issue they were encountering in `exploit/windows/mysql/mysql_mof`, whereby they couldn't set a blank password.  In chasing down their issue, I realized that you'd already identified the same cause.

They are trying to authenticate using an empty string as a password, but validation was preventing them from running.  Note that both single and double quotes failed, even though `irb` reported that `datastore['PASSWORD']` was not `nil`.

Injecting the `pry` debugger into `OptString.valid?` shows that `super` is being returned, which is set to `false`:

```
From: /home/administrator/git/r7/metasploit-framework/lib/msf/core/opt_string.rb @ line 32 Msf::OptString#valid?:

    31: def valid?(value=self.value, check_empty: true)
 => 32:   require 'pry'; binding.pry
    33:   value = normalize(value)
    34:   return false if check_empty && empty_required_value?(value)
    35:   return super
    36: end

[1] pry(#<Msf::OptString>)> name
=> "PASSWORD"
[2] pry(#<Msf::OptString>)> value
=> ""
[3] pry(#<Msf::OptString>)> check_empty
=> true
[4] pry(#<Msf::OptString>)> empty_required_value?(value)
=> false
[5] pry(#<Msf::OptString>)> super
=> false
```

I would like to propose that we modify the final line of `OptString.valid?`' to `return true`:

https://github.com/rapid7/metasploit-framework/blob/cf9bf5934f78720cc8a92f9046c4dc8c8f4a430b/lib/msf/core/opt_string.rb#L31-L35

To get the ball rolling, I'll get a PR up and start testing, since I wouldn't be surprised if this has additional repercussions.  Feedback welcome.  😅

**Comment by bcoles** [comment]
Pretty sure this was fixed by #11061. Closing.
